File: 0368-largest-divisible-subset/0368-largest-divisible-subset.kt
class Solution {
    fun largestDivisibleSubset(nums: IntArray): List<Int> {
        val sorted = nums.sorted()
        val d = IntArray(nums.size)
        val prev = IntArray(nums.size) { -1 }
        d[0] = 1
        
        var index = 0
        for (i in 1..sorted.lastIndex) {
            val num = sorted[i]
            for (j in 0..< i) {
                if (num % sorted[j] != 0) {
                    continue
                }

                if (d[i] < d[j]) {
                    d[i] = d[j]
                    prev[i] = j
                }                
            }

            d[i]++
            if (d[index] < d[i]) {
                index = i
            }
        }

        val answer = mutableListOf<Int>()
        while (index != -1) {
            answer += sorted[index]
            index = prev[index]
        }

        return answer
    }
}

File: 0368-largest-divisible-subset/0368-largest-divisible-subset.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        max_count = 0
        max_index = 0

        dp = [0] * len(nums)
        prev = list(range(len(nums))) 
        sorted_nums = sorted(nums)
        for i, num in enumerate(sorted_nums):
            for j in range(i):
                if num % sorted_nums[j] != 0:
                    continue

                if dp[i] < dp[j]:
                    dp[i] = dp[j]
                    prev[i] = j

            dp[i] += 1
            if max_count < dp[i]:
                max_count = dp[i]
                max_index = i

        answer = [sorted_nums[max_index]]
        while max_index != prev[max_index]:
            max_index = prev[max_index]
            answer.append(sorted_nums[max_index])

        return answer